# T05 Motor trend Car Road Test

|                |   |
:----------------|---|
| **Nombre**     |Maria Fernanda Muñoz Sevilla   |
| **Fecha**      |15/09/25   |
| **Expediente** |751190   |

Utiliza el archivo "Motor Trend Car Road Tests.xlsx" y completa las siguientes actividades:

## 1.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Considera todos los demás factores como numéricos/ordinales.

In [2]:
import pandas as pd
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression, Ridge, Lasso
import numpy as np
import matplotlib.pyplot as plt
from sklearn.metrics import r2_score
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

In [3]:
df=pd.read_excel("C:/Users/munoz/Downloads/lab_apre_est/Motor Trend Car Road Tests.xlsx")
df.head()

model   mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  \
0          Mazda RX4  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4   
1      Mazda RX4 Wag  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4   
2         Datsun 710  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4   
3     Hornet 4 Drive  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3   
4  Hornet Sportabout  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3   

   carb  
0     4  
1     4  
2     1  
3     1  
4     2

In [4]:
dat=df.drop(columns=["model"])
dat.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8  360.0  175  3.15  3.440  17.02   0   0     3     2

- Calcula el R2 e interpreta los signos de los betas.

In [9]:
x=dat.iloc[:,1:].values
y=dat.mpg
n=len(y)
unos=np.ones([n,1])
X=np.hstack([unos,x])
ols=sm.OLS(y,X)
resultados=ols.fit()
resultados.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.869
Model:                            OLS   Adj. R-squared:                  0.807
Method:                 Least Squares   F-statistic:                     13.93
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           3.79e-07
Time:                        07:11:25   Log-Likelihood:                -69.855
No. Observations:                  32   AIC:                             161.7
Df Residuals:                      21   BIC:                             177.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         12.3034     18.718      0.657      0.518     -26.623      51.229
x1            -0.1114      1.045     -0.107      0.916      -2.285       2.062
x2             0.0133      0.018      0.747      0.463      -0.024       0.050
x3            -0.0215      0.022     -0.987      0.335      -0.067       0.024
x4             0.7871      1.635      0.481      0.635      -2.614       4.188
x5            -3.7153      1.894     -1.961      0.063      -7.655       0.224
x6             0.8210      0.731      1.123      0.274      -0.699       2.341
x7             0.3178      2.105      0.151      0.881      -4.059       4.694
x8             2.5202      2.057      1.225      0.234      -1.757       6.797
x9             0.6554      1.493      0.439      0.665      -2.450       3.761
x10           -0.1994      0.829     -0.241      0.812      -1.923       1.524
==============================================================================
Omnibus:                        1.907   Durbin-Watson:                   1.861
Prob(Omnibus):                  0.385   Jarque-Bera (JB):                1.747
Skew:                           0.521   Prob(JB):                        0.418
Kurtosis:                       2.526   Cond. No.                     1.22e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.22e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

En esta regresión lineal sin escalar ni entrenar, podemos observar que tenemos betas tanto por debajo de cero, como muy por encima. Los betas que se encuentran por encima de cero, nos hablan de una relación donde incrementa nuestra variable y, mientras que, las betas menores a 0 representan la disminución en y. Además, nuestra r2 nos dice que de nuestro modelo se puede explicar el 86.9%, mientra que el resto queda en el error.


- Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [13]:
#separamos los datos
x_train1, x_test1, y_train1, y_test1 = train_test_split(X, y, test_size=0.4, random_state=137)

#estandarizamos
scaler=StandardScaler().fit(x_train1)

x_train_scaled1=scaler.transform(x_train1)
x_test_scaled1=scaler.transform(x_test1)

In [15]:
#Hacemos la regresión lineal
x_train_e= sm.add_constant(x_train_scaled1)
resultado= sm.OLS(y_train1,x_train_e).fit()
resultado.summary()

C:\Users\munoz\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=19
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.879
Model:                            OLS   Adj. R-squared:                  0.727
Method:                 Least Squares   F-statistic:                     5.802
Date:                Thu, 18 Sep 2025   Prob (F-statistic):             0.0101
Time:                        07:11:27   Log-Likelihood:                -40.712
No. Observations:                  19   AIC:                             103.4
Df Residuals:                       8   BIC:                             113.8
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.4947      0.729     25.366      0.000      16.813      20.176
x1          1.182e-15   1.98e-15      0.598      0.566   -3.37e-15    5.74e-15
x2            -2.8779      3.427     -0.840      0.425     -10.781       5.025
x3            -0.9201      3.625     -0.254      0.806      -9.280       7.440
x4             0.1825      2.839      0.064      0.950      -6.363       6.728
x5             0.2609      1.492      0.175      0.865      -3.179       3.701
x6            -2.1215      3.207     -0.662      0.527      -9.517       5.274
x7             0.1404      1.998      0.070      0.946      -4.468       4.749
x8            -0.6393      2.275     -0.281      0.786      -5.884       4.606
x9             0.3812      1.705      0.224      0.829      -3.550       4.312
x10           -0.1370      2.167     -0.063      0.951      -5.134       4.860
x11           -0.6824      2.324     -0.294      0.776      -6.041       4.676
==============================================================================
Omnibus:                        3.180   Durbin-Watson:                   1.104
Prob(Omnibus):                  0.204   Jarque-Bera (JB):                1.324
Skew:                           0.480   Prob(JB):                        0.516
Kurtosis:                       3.867   Cond. No.                     3.21e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.02e-31. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Ahora que escalamos podemos ver que incrementaron nuestras betas negativas, dandonos a entender que en este modelo la mayoria de nuestras variables hacen que disminuya nuestra variable y ("mpg"). También, observamos que nuestro r2 es de 0.879, por lo que, el 87.9% es explicado por el modelo, el resto queda del error.

In [17]:
#Sacamos lor R2 de x_train y x_test

#preparamos x_test
x_test_e=sm.add_constant(x_test_scaled1)
y_pred1=resultado.predict(x_test_e)
#calculamos r2 de xtrain
r2_train=resultado.rsquared
print("R2 de train: ",r2_train)

#calculamos r2 de xtest
r2_test=r2_score(y_test1, y_pred1)
print("R2 de test: ",r2_test)

R2 de train:  0.8788249154881685
R2 de test:  0.6781478185882241


- Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [20]:
#Con alfa = 0.5, con xtrain_scaled y xtest_scaled
ridge1= Ridge(alpha=0.5)
ridge1.fit(x_train_scaled1, y_train1)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_train1=ridge1.predict(x_train_scaled1)
y_pred_test1=ridge1.predict(x_test_scaled1)

#sacamos r2
r2_ridge_train1=r2_score(y_train1,y_pred_train1)
r2_ridge_test1=r2_score(y_test1,y_pred_test1)
print("R2 de train: ", r2_ridge_train1)
print("R2 de test: ", r2_ridge_test1)



R2 de train:  0.8773082539790075
R2 de test:  0.7170867292018437


In [22]:
#Con alfa=1
ridge2= Ridge(alpha=1)
ridge2.fit(x_train_scaled1, y_train1)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_train2=ridge2.predict(x_train_scaled1)
y_pred_test2=ridge2.predict(x_test_scaled1)

#sacamos r2
r2_ridge_train2=r2_score(y_train1,y_pred_train2)
r2_ridge_test2=r2_score(y_test1,y_pred_test2)
print("R2 de train: ", r2_ridge_train2)
print("R2 de test: ", r2_ridge_test2)

R2 de train:  0.8757139730165581
R2 de test:  0.7282860802849515


In [24]:
# Con alfa=10
ridge3= Ridge(alpha=10)
ridge3.fit(x_train_scaled1, y_train1)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_train3=ridge3.predict(x_train_scaled1)
y_pred_test3=ridge3.predict(x_test_scaled1)

#sacamos r2
r2_ridge_train3=r2_score(y_train1,y_pred_train3)
r2_ridge_test3=r2_score(y_test1,y_pred_test3)
print("R2 de train: ", r2_ridge_train3)
print("R2 de test: ", r2_ridge_test3)

R2 de train:  0.8601362367025545
R2 de test:  0.7431630893996333


Vinedo los resultados, considero que el mejor ajuste fue el de lambda = 10, ya que, muestra una mayor generalización de los datos. Mientras que, para lambda = 1, es un mejor balance, y el lambda = 0.5, es muy parecido al modelo sin penalizar, ya que el ajuste es muy pequeño.

## 1.2 Repite el ejercicio anterior usando 'qsec' como salida.

- Calcula el R2 e interpreta los signos de los betas.

In [28]:
x_12=dat.drop(columns=["qsec"]).values.reshape(-1,10)
y_12=dat.qsec
n=len(y_12)
unos=np.ones([n,1])
X_12=np.hstack([unos,x_12])
ols=sm.OLS(y_12,X_12)
resultados=ols.fit()
resultados.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.815
Method:                 Least Squares   F-statistic:                     14.66
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           2.44e-07
Time:                        07:11:36   Log-Likelihood:                -30.242
No. Observations:                  32   AIC:                             82.48
Df Residuals:                      21   BIC:                             98.61
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.7762      3.876      4.586      0.000       9.716      25.837
x1             0.0690      0.061      1.123      0.274      -0.059       0.197
x2            -0.3627      0.293     -1.239      0.229      -0.971       0.246
x3            -0.0075      0.005     -1.505      0.147      -0.018       0.003
x4            -0.0016      0.006     -0.242      0.811      -0.015       0.012
x5            -0.1311      0.476     -0.275      0.786      -1.121       0.859
x6             1.4963      0.500      2.990      0.007       0.456       2.537
x7             0.9700      0.573      1.694      0.105      -0.221       2.161
x8            -0.9012      0.585     -1.540      0.139      -2.118       0.316
x9            -0.2013      0.433     -0.465      0.647      -1.101       0.699
x10           -0.2736      0.233     -1.174      0.254      -0.758       0.211
==============================================================================
Omnibus:                       21.069   Durbin-Watson:                   2.573
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               38.291
Skew:                           1.470   Prob(JB):                     4.84e-09
Kurtosis:                       7.481   Cond. No.                     8.77e+03
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.77e+03. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Podmeos observar que el modelo presenta más variables que disminuyen nuestra variable y ("qsec"), además, el modelo puede explicar el 87.5%, muy parecido al modelo anterior con y=mpg.

- Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [31]:
#separamos los datos
x_train12, x_test12, y_train12, y_test12 = train_test_split(X_12, y_12, test_size=0.4, random_state=137)

#estandarizamos
scaler=StandardScaler().fit(x_train12)

x_train_scaled12=scaler.transform(x_train12)
x_test_scaled12=scaler.transform(x_test12)

In [33]:
#Hacemos la regresión lineal
x_train_e12= sm.add_constant(x_train_scaled12)
resultado= sm.OLS(y_train12,x_train_e12).fit()
resultado.summary()

C:\Users\munoz\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=19
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.867
Model:                            OLS   Adj. R-squared:                  0.701
Method:                 Least Squares   F-statistic:                     5.214
Date:                Thu, 18 Sep 2025   Prob (F-statistic):             0.0140
Time:                        07:11:38   Log-Likelihood:                -20.168
No. Observations:                  19   AIC:                             62.34
Df Residuals:                       8   BIC:                             72.73
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8511      0.247     72.185      0.000      17.281      18.421
x1          4.826e-16   8.01e-16      0.603      0.563   -1.36e-15    2.33e-15
x2             0.0499      0.710      0.070      0.946      -1.588       1.688
x3            -1.5136      1.088     -1.391      0.202      -4.023       0.995
x4            -1.3262      1.142     -1.161      0.279      -3.960       1.307
x5             0.5221      0.945      0.552      0.596      -1.657       2.702
x6             0.0714      0.506      0.141      0.891      -1.096       1.239
x7             1.5030      0.983      1.530      0.165      -0.763       3.769
x8            -0.1762      0.773     -0.228      0.825      -1.958       1.606
x9            -0.6846      0.527     -1.299      0.230      -1.900       0.531
x10           -0.1163      0.734     -0.158      0.878      -1.809       1.576
x11           -0.7713      0.744     -1.037      0.330      -2.487       0.944
==============================================================================
Omnibus:                        0.980   Durbin-Watson:                   2.087
Prob(Omnibus):                  0.613   Jarque-Bera (JB):                0.756
Skew:                           0.453   Prob(JB):                        0.685
Kurtosis:                       2.634   Cond. No.                     1.07e+17
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 1.02e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

AL escalar, entrenar y sacar la prueba podemos ver que el r2 dismunuyo a 0.867, esto porque tenemos un mejor ajuste de los datos. De igual forma, la mayoría de las variables presentan una disminución con la relación de la variable y

In [35]:
#Sacamos lor R2 de x_train y x_test

#preparamos x_test
x_test_e12=sm.add_constant(x_test_scaled12)
y_pred12=resultado.predict(x_test_e12)
#calculamos r2 de xtrain
r2_train12=resultado.rsquared
print("R2 de train: ",r2_train12)

#calculamos r2 de xtest
r2_test12=r2_score(y_test12, y_pred12)
print("R2 de test: ",r2_test12)

R2 de train:  0.8669774049790414
R2 de test:  0.6814395697520064


- Añade regularización L2 con un hiperparámetro lambda decidido por ti. Cambia este valor y compara con varios distintos los R2 de entrenamiento y de prueba.

In [38]:
#Con alfa = 0.5, con xtrain_scaled y xtest_scaled
ridge12= Ridge(alpha=0.5)
ridge12.fit(x_train_scaled12, y_train12)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_train12=ridge12.predict(x_train_scaled12)
y_pred_test12=ridge12.predict(x_test_scaled12)

#sacamos r2
r2_ridge_train12=r2_score(y_train12,y_pred_train12)
r2_ridge_test12=r2_score(y_test12,y_pred_test12)
print("R2 de train: ", r2_ridge_train12)
print("R2 de test: ", r2_ridge_test12)



R2 de train:  0.8566464863628895
R2 de test:  0.8054592665804765


In [40]:
#Con alfa = 1, con xtrain_scaled y xtest_scaled
ridge12_2= Ridge(alpha=1)
ridge12_2.fit(x_train_scaled12, y_train12)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_train12_2=ridge12_2.predict(x_train_scaled12)
y_pred_test12_2=ridge12_2.predict(x_test_scaled12)

#sacamos r2
r2_ridge_train12_2=r2_score(y_train12,y_pred_train12_2)
r2_ridge_test12_2=r2_score(y_test12,y_pred_test12_2)
print("R2 de train: ", r2_ridge_train12_2)
print("R2 de test: ", r2_ridge_test12_2)

R2 de train:  0.8483218557627674
R2 de test:  0.8227910634318731


In [42]:
#Con alfa = 10, con xtrain_scaled y xtest_scaled
ridge12_3= Ridge(alpha=10)
ridge12_3.fit(x_train_scaled12, y_train12)
#preparamos para sacar las r2 de ridge1 (nuestro modelo), sacando las y_pred, 
y_pred_train12_3=ridge12_3.predict(x_train_scaled12)
y_pred_test12_3=ridge12_3.predict(x_test_scaled12)

#sacamos r2
r2_ridge_train12_3=r2_score(y_train12,y_pred_train12_3)
r2_ridge_test12_3=r2_score(y_test12,y_pred_test12_3)
print("R2 de train: ", r2_ridge_train12_3)
print("R2 de test: ", r2_ridge_test12_3)

R2 de train:  0.772799934397899
R2 de test:  0.8030251739579517


Al observar las r2, considero que la mejor es la de lambda = 1, ya que, es la que presenta una mejor generalización de las 3.

## 2.1 Realiza una regresión tomando 'mpg' como salida y eliminando la columna 'model'. Crea columnas dummies para los factores 'cyl', 'gear' y 'carb'.

In [45]:
dat_dumm=pd.get_dummies(dat, columns=["cyl","gear","carb"])
dat_dumm

mpg   disp   hp  drat     wt   qsec  vs  am  cyl_4  cyl_6  cyl_8  gear_3  \
0   21.0  160.0  110  3.90  2.620  16.46   0   1  False   True  False   False   
1   21.0  160.0  110  3.90  2.875  17.02   0   1  False   True  False   False   
2   22.8  108.0   93  3.85  2.320  18.61   1   1   True  False  False   False   
3   21.4  258.0  110  3.08  3.215  19.44   1   0  False   True  False    True   
4   18.7  360.0  175  3.15  3.440  17.02   0   0  False  False   True    True   
5   18.1  225.0  105  2.76  3.460  20.22   1   0  False   True  False    True   
6   14.3  360.0  245  3.21  3.570  15.84   0   0  False  False   True    True   
7   24.4  146.7   62  3.69  3.190  20.00   1   0   True  False  False   False   
8   22.8  140.8   95  3.92  3.150  22.90   1   0   True  False  False   False   
9   19.2  167.6  123  3.92  3.440  18.30   1   0  False   True  False   False   
10  17.8  167.6  123  3.92  3.440  18.90   1   0  False   True  False   False   
11  16.4  275.8  180  3.07  4.070  17.40   0   0  False  False   True    True   
12  17.3  275.8  180  3.07  3.730  17.60   0   0  False  False   True    True   
13  15.2  275.8  180  3.07  3.780  18.00   0   0  False  False   True    True   
14  10.4  472.0  205  2.93  5.250  17.98   0   0  False  False   True    True   
15  10.4  460.0  215  3.00  5.424  17.82   0   0  False  False   True    True   
16  14.7  440.0  230  3.23  5.345  17.42   0   0  False  False   True    True   
17  32.4   78.7   66  4.08  2.200  19.47   1   1   True  False  False   False   
18  30.4   75.7   52  4.93  1.615  18.52   1   1   True  False  False   False   
19  33.9   71.1   65  4.22  1.835  19.90   1   1   True  False  False   False   
20  21.5  120.1   97  3.70  2.465  20.01   1   0   True  False  False    True   
21  15.5  318.0  150  2.76  3.520  16.87   0   0  False  False   True    True   
22  15.2  304.0  150  3.15  3.435  17.30   0   0  False  False   True    True   
23  13.3  350.0  245  3.73  3.840  15.41   0   0  False  False   True    True   
24  19.2  400.0  175  3.08  3.845  17.05   0   0  False  False   True    True   
25  27.3   79.0   66  4.08  1.935  18.90   1   1   True  False  False   False   
26  26.0  120.3   91  4.43  2.140  16.70   0   1   True  False  False   False   
27  30.4   95.1  113  3.77  1.513  16.90   1   1   True  False  False   False   
28  15.8  351.0  264  4.22  3.170  14.50   0   1  False  False   True   False   
29  19.7  145.0  175  3.62  2.770  15.50   0   1  False   True  False   False   
30  15.0  301.0  335  3.54  3.570  14.60   0   1  False  False   True   False   
31  21.4  121.0  109  4.11  2.780  18.60   1   1   True  False  False   False   

    gear_4  gear_5  carb_1  carb_2  carb_3  carb_4  carb_6  carb_8  
0     True   False   False   False   False    True   False   False  
1     True   False   False   False   False    True   False   False  
2     True   False    True   False   False   False   False   False  
3    False   False    True   False   False   False   False   False  
4    False   False   False    True   False   False   False   False  
5    False   False    True   False   False   False   False   False  
6    False   False   False   False   False    True   False   False  
7     True   False   False    True   False   False   False   False  
8     True   False   False    True   False   False   False   False  
9     True   False   False   False   False    True   False   False  
10    True   False   False   False   False    True   False   False  
11   False   False   False   False    True   False   False   False  
12   False   False   False   False    True   False   False   False  
13   False   False   False   False    True   False   False   False  
14   False   False   False   False   False    True   False   False  
15   False   False   False   False   False    True   False   False  
16   False   False   False   False   False    True   False   False  
17    True   False    True   False   False   False   False   False  
18    True   False  

- Calcula el R2 e interpreta los signos de los betas.

In [48]:
x2=dat_dumm.iloc[:,1:].astype(float).values
y2=dat_dumm.mpg
n=len(y2)
unos=np.ones([n,1])
X2=np.hstack([unos,x2])
ols=sm.OLS(y2,X2)
resultados=ols.fit()
resultados.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.893
Model:                            OLS   Adj. R-squared:                  0.779
Method:                 Least Squares   F-statistic:                     7.830
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           0.000124
Time:                        07:11:42   Log-Likelihood:                -66.608
No. Observations:                  32   AIC:                             167.2
Df Residuals:                      15   BIC:                             192.1
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         14.4937     10.672      1.358      0.195      -8.254      37.241
x1             0.0355      0.032      1.114      0.283      -0.032       0.104
x2            -0.0705      0.039     -1.788      0.094      -0.155       0.014
x3             1.1828      2.483      0.476      0.641      -4.111       6.476
x4            -4.5298      2.539     -1.784      0.095      -9.941       0.881
x5             0.3678      0.935      0.393      0.700      -1.626       2.362
x6             1.9309      2.871      0.672      0.512      -4.189       8.051
x7             1.2121      3.214      0.377      0.711      -5.637       8.062
x8             5.8262      5.651      1.031      0.319      -6.218      17.871
x9             3.1775      3.526      0.901      0.382      -4.337      10.692
x10            5.4900      4.884      1.124      0.279      -4.920      15.900
x11            3.6170      3.548      1.020      0.324      -3.945      11.179
x12            4.7313      4.622      1.024      0.322      -5.120      14.583
x13            6.1454      4.099      1.499      0.155      -2.591      14.882
x14           -0.0577      3.844     -0.015      0.988      -8.250       8.135
x15           -1.0370      2.923     -0.355      0.728      -7.267       5.193
x16            2.9420      3.393      0.867      0.400      -4.291      10.175
x17            1.0338      2.628      0.393      0.700      -4.567       6.634
x18            4.4199      4.041      1.094      0.291      -4.193      13.033
x19            7.1927      5.924      1.214      0.243      -5.434      19.820
==============================================================================
Omnibus:                        2.468   Durbin-Watson:                   2.105
Prob(Omnibus):                  0.291   Jarque-Bera (JB):                1.772
Skew:                           0.576   Prob(JB):                        0.412
Kurtosis:                       3.009   Cond. No.                     1.21e+19
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 2.04e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

En esta regresión linea sin escalar ni entrenar, donde convertimos a dummies parte de nuestros datos, podemos ver que nuestro modelo presenta más variables que se relación con el incremento de mpg. Además, de tener un r2 de 0.893, presentando lo que se puede explicar mediante el modelo.

- Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba

In [51]:
#separamos los datos
x_train2, x_test2, y_train2, y_test2 = train_test_split(X2, y2, test_size=0.4, random_state=137)

#escalamos los xtrain y xtest
scaler=StandardScaler().fit(x_train2)

x_train_scaled2=scaler.transform(x_train2)
x_test_scaled2=scaler.transform(x_test2)


In [53]:
#Hacemos la regresión lineal
x_train_e2= sm.add_constant(x_train_scaled2)
resultado= sm.OLS(y_train2,x_train_e2).fit()
resultado.summary()

C:\Users\munoz\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=19
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.986
Model:                            OLS   Adj. R-squared:                  0.916
Method:                 Least Squares   F-statistic:                     14.15
Date:                Thu, 18 Sep 2025   Prob (F-statistic):             0.0253
Time:                        07:11:43   Log-Likelihood:                -20.169
No. Observations:                  19   AIC:                             72.34
Df Residuals:                       3   BIC:                             87.45
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         18.4947      0.404     45.797      0.000      17.210      19.780
x1         -2.234e-14   1.96e-14     -1.141      0.337   -8.46e-14       4e-14
x2           -14.4986      7.624     -1.902      0.153     -38.762       9.765
x3            -3.3690      3.213     -1.049      0.371     -13.594       6.856
x4            -0.1669      3.561     -0.047      0.966     -11.499      11.165
x5             4.4120      3.492      1.263      0.296      -6.702      15.526
x6            -0.5129      1.731     -0.296      0.786      -6.022       4.996
x7            -2.3637      2.134     -1.108      0.349      -9.154       4.426
x8            -1.0955      1.784     -0.614      0.583      -6.773       4.582
x9             0.7602      2.173      0.350      0.750      -6.155       7.675
x10           -3.3021      1.507     -2.191      0.116      -8.099       1.494
x11            1.7608      2.908      0.605      0.588      -7.495      11.016
x12            0.4611      1.985      0.232      0.831      -5.857       6.779
x13           -2.2382      1.657     -1.350      0.270      -7.512       3.036
x14            2.2217      1.152      1.928      0.149      -1.445       5.889
x15            1.5335      1.309      1.172      0.326      -2.632       5.699
x16           -1.7334      0.777     -2.231      0.112      -4.206       0.739
x17           -2.5749      1.361     -1.891      0.155      -6.907       1.758
x18            2.6488      1.456      1.819      0.166      -1.985       7.283
x19                 0          0        nan        nan           0           0
x20           -0.0164      2.228     -0.007      0.995      -7.108       7.075
==============================================================================
Omnibus:                        2.711   Durbin-Watson:                   1.924
Prob(Omnibus):                  0.258   Jarque-Bera (JB):                1.095
Skew:                           0.503   Prob(JB):                        0.578
Kurtosis:                       3.610   Cond. No.                     6.70e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
[3] The smallest eigenvalue is 3.19e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Al realizar el escalamiento y sacar las x_train y x_test, vemos que nuestra r2 incremento a un 0.986, lo cual nos indica el buen escalamiento que tuvo nuestro modelo, pues explica casi el modelo completo. Así mismo, tuvimos variables que ahora con su relación con y no es en aumento, sino en disminución.

In [55]:
#Sacamos lor R2 de x_train y x_test

#preparamos x_test
x_test_e2=sm.add_constant(x_test_scaled2, has_constant='add')
y_pred2=resultado.predict(x_test_e2)
#calculamos r2 de xtrain
r2_train2=resultado.rsquared
print("R2 de train: ",r2_train2)

#calculamos r2 de xtest
r2_test2=r2_score(y_test2, y_pred2)
print("R2 de test: ",r2_test2)

R2 de train:  0.9860599593190703
R2 de test:  -1.7261974546344292


Sacando las r2 de train y test, podemos ver que hay overfitting, por lo que, el modelo no es tan bueno como se había presentado en la regresión lineal.

## 2.2 Repite el ejercicio anterior usando 'qsec' como salida.

- Calcula el R2 e interpreta los signos de los betas.

In [59]:
x_22=dat_dumm.drop(columns=["qsec"]).astype(float).values.reshape(-1,19)
y_22=dat_dumm.qsec
n=len(y_22)
unos=np.ones([n,1])
X_22=np.hstack([unos,x_22])
ols=sm.OLS(y_22,X_22)
resultados=ols.fit()
resultados.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.908
Model:                            OLS   Adj. R-squared:                  0.810
Method:                 Least Squares   F-statistic:                     9.283
Date:                Thu, 18 Sep 2025   Prob (F-statistic):           4.35e-05
Time:                        07:11:45   Log-Likelihood:                -25.252
No. Observations:                  32   AIC:                             84.50
Df Residuals:                      15   BIC:                             109.4
Df Model:                          16                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          7.9157      2.339      3.385      0.004       2.931      12.900
x1             0.0277      0.071      0.393      0.700      -0.123       0.178
x2             0.0035      0.009      0.389      0.703      -0.016       0.023
x3            -0.0021      0.012     -0.173      0.865      -0.027       0.023
x4             0.1079      0.687      0.157      0.877      -1.356       1.571
x5             0.8105      0.739      1.097      0.290      -0.764       2.385
x6             0.2657      0.797      0.333      0.744      -1.434       1.965
x7            -1.6943      0.771     -2.197      0.044      -3.338      -0.050
x8             3.9957      1.231      3.247      0.005       1.373       6.619
x9             2.8913      0.656      4.405      0.001       1.492       4.290
x10            1.0288      1.371      0.750      0.465      -1.894       3.951
x11            2.1337      0.843      2.530      0.023       0.336       3.931
x12            3.4660      0.961      3.608      0.003       1.418       5.514
x13            2.3160      1.049      2.209      0.043       0.081       4.551
x14            2.3040      0.872      2.642      0.018       0.445       4.163
x15            1.4696      0.711      2.066      0.057      -0.046       2.985
x16            2.0747      0.791      2.624      0.019       0.390       3.760
x17            0.3570      0.719      0.496      0.627      -1.176       1.890
x18            0.7388      1.137      0.650      0.526      -1.685       3.163
x19            0.9717      1.686      0.576      0.573      -2.623       4.566
==============================================================================
Omnibus:                       12.619   Durbin-Watson:                   2.588
Prob(Omnibus):                  0.002   Jarque-Bera (JB):               16.989
Skew:                           0.924   Prob(JB):                     0.000205
Kurtosis:                       6.054   Cond. No.                     7.46e+18
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The smallest eigenvalue is 5.33e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

En este modelo sin escalar ni entrenar, vemos que tenemos un r2 de 0.908, lo cual es bueno porque explica la gran mayoría de los datos. También vemos que la relación entre las variables con y en mayormente positiva

- Realiza un train-test-split donde se use el 40% de los datos para entrenar. Calcula el R2 de entrenamiento y de prueba.

In [62]:
#separamos los datos
x_train22, x_test22, y_train22, y_test22 = train_test_split(X_22, y_22, test_size=0.4, random_state=137)

#escalamos los xtrain y xtest
scaler=StandardScaler().fit(x_train22)

x_train_scaled22=scaler.transform(x_train22)
x_test_scaled22=scaler.transform(x_test22)


In [84]:
#Hacemos la regresión lineal
x_train_e22= sm.add_constant(x_train_scaled22)
resultado= sm.OLS(y_train22,x_train_e22).fit()
resultado.summary()

C:\Users\munoz\anaconda3\Lib\site-packages\scipy\stats\_axis_nan_policy.py:531: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=19
  res = hypotest_fun_out(*samples, **kwds)


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   qsec   R-squared:                       0.947
Model:                            OLS   Adj. R-squared:                  0.683
Method:                 Least Squares   F-statistic:                     3.583
Date:                Thu, 18 Sep 2025   Prob (F-statistic):              0.160
Time:                        08:10:31   Log-Likelihood:                -11.403
No. Observations:                  19   AIC:                             54.81
Df Residuals:                       3   BIC:                             69.92
Df Model:                          15                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.8511      0.255     70.116      0.000      17.041      18.661
x1         -3.193e-16   1.15e-14     -0.028      0.980   -3.69e-14    3.62e-14
x2            -0.6298      2.125     -0.296      0.786      -7.394       6.134
x3             1.1599      7.107      0.163      0.881     -21.456      23.776
x4            -0.5011      2.350     -0.213      0.845      -7.980       6.978
x5             1.4781      2.077      0.712      0.528      -5.133       8.089
x6             0.2834      2.720      0.104      0.924      -8.373       8.940
x7            -1.2929      1.411     -0.916      0.427      -5.785       3.199
x8            -1.5616      0.782     -1.998      0.140      -4.049       0.926
x9             1.7809      0.946      1.882      0.156      -1.231       4.793
x10            0.4684      1.508      0.311      0.776      -4.331       5.268
x11           -1.9343      1.589     -1.217      0.311      -6.991       3.123
x12            0.2972      1.251      0.238      0.828      -3.684       4.279
x13           -0.2901      1.314     -0.221      0.839      -4.473       3.892
x14           -0.0371      1.087     -0.034      0.975      -3.496       3.422
x15            0.6935      0.912      0.760      0.502      -2.210       3.596
x16           -0.0302      0.799     -0.038      0.972      -2.571       2.511
x17            0.2974      1.259      0.236      0.829      -3.710       4.305
x18           -0.7588      1.257     -0.604      0.589      -4.759       3.242
x19                 0          0        nan        nan           0           0
x20            0.2633      1.397      0.189      0.862      -4.181       4.708
==============================================================================
Omnibus:                        6.980   Durbin-Watson:                   2.414
Prob(Omnibus):                  0.030   Jarque-Bera (JB):                5.808
Skew:                          -0.482   Prob(JB):                       0.0548
Kurtosis:                       5.531   Cond. No.                     9.75e+16
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The input rank is higher than the number of observations.
[3] The smallest eigenvalue is 1.61e-32. This might indicate that there are
strong multicollinearity problems or that the design matrix is singular.
"""

Al escalar y entrenar nuestro modelo, vemos que nuestro r2 incremento a 0.947, además ahora presentamos más variables con relación negativa con nuestra y

In [66]:
#Sacamos lor R2 de x_train y x_test

#preparamos x_test
x_test_e22=sm.add_constant(x_test_scaled22,  has_constant='add')
y_pred22=resultado.predict(x_test_e22)
#calculamos r2 de xtrain
r2_train22=resultado.rsquared
print("R2 de train: ",r2_train22)

#calculamos r2 de xtest
r2_test22=r2_score(y_test22, y_pred22)
print("R2 de test: ",r2_test22)

R2 de train:  0.9471283557499286
R2 de test:  -1.5356274676070791


Sacando las r2 de train y test, podemos ver que hay overfitting, por lo que, el modelo no es tan bueno como se había presentado en la regresión lineal.

## 3.1 Compara los R2 de los ejercicios 1.1 & 2.1.

In [69]:
print("R2 del ejercicio 1.1 sin dummies: ",r2_train)
print("R2 del ejercicio 2.1 con dummies: ",r2_train2)

R2 del ejercicio 1.1 sin dummies:  0.8788249154881685
R2 del ejercicio 2.1 con dummies:  0.9860599593190703


Observando estar r2, diriamos que es mejor el modelo con dummies, pero este queda descartado cuando observamos su r2_test, el cual salio negativo

## 3.2 Compara los R2 de los ejercicios 1.2 & 2.2.

In [72]:
print("R2 del ejercicio 1.2 sin dummies: ",r2_train12)
print("R2 del ejercicio 2.2 con dummies: ",r2_train22)

R2 del ejercicio 1.2 sin dummies:  0.8669774049790414
R2 del ejercicio 2.2 con dummies:  0.9471283557499286


Así como la comparación anterior, pasa en esta, pues vemos que el r2 de las dummies es mayor y que explica más el modelo, pero cuando observamos el r2 de test, vemos que hay overfitting